homeassistant/components/fan/reproduce_state.py

```diff
@@ -56,37 +61,51 @@ async def _async_reproduce_state(
-
-    service_data = {ATTR_ENTITY_ID: state.entity_id}
-    service_calls = {}  # service: service_data
+    service_calls: list[tuple[str, dict[str, Any]]] = []
 
     if state.state == STATE_ON:
         # The fan should be on
         if cur_state.state != STATE_ON:
-            # Turn on the fan at first
-            service_calls[SERVICE_TURN_ON] = service_data
-
-        for attr, service in ATTRIBUTES.items():
-            # Call services to adjust the attributes
-            if attr in state.attributes and not check_attr_equal(
-                state.attributes, cur_state.attributes, attr
-            ):
-                data = service_data.copy()
-                data[attr] = state.attributes[attr]
-                service_calls[service] = data
-
-    elif state.state == STATE_OFF:
-        service_calls[SERVICE_TURN_OFF] = service_data
-
-    for service, data in service_calls.items():
+            # Turn on the fan with all the speed and modes settings.
+            # The `turn_on` method will figure out in which mode to
+            # turn the fan on.
+            service_calls.append(
+                (
+                    SERVICE_TURN_ON,
+                    {
+                        attr: state.attributes.get(attr)
+                        for attr in SPEED_AND_MODES_ATTRIBUTES
+                        if state.attributes.get(attr) is not None
+                    },
+                )
+            )
+        else:
+            # If the fan is already on, we need to set speed or mode
+            # based on the state.
+            #
+            # Speed and preset mode are mutually exclusive, so one of
+            # them is always going to be stored as None. If we were to
+            # try to set it, it will raise an error. So instead we
+            # only update the one that is non-None.
+            for attr, service in SPEED_AND_MODES_ATTRIBUTES.items():
+                value = state.attributes.get(attr)
+                if value is not None and value != cur_state.attributes.get(attr):
+                    service_calls.append((service, {attr: value}))
+
+        # The simple attributes are copied directly. They can only be
+        # None if the fan does not support the feature in the first
+        # place, so the equality check ensures we don't call the
+        # services with invalid parameters.
+        for attr, service in SIMPLE_ATTRIBUTES.items():
+            if (value := state.attributes.get(attr)) != cur_state.attributes.get(attr):
+                service_calls.append((service, {attr: value}))
+    elif state.state == STATE_OFF and cur_state.state != state.state:
+        service_calls.append((SERVICE_TURN_OFF, {}))
+
+    service_data = {ATTR_ENTITY_ID: state.entity_id}
+    for service, data in service_calls:
         await hass.services.async_call(
-            DOMAIN, service, data, context=context, blocking=True
+            DOMAIN, service, dict(data, **service_data), context=context, blocking=True
```

Comment:
Optional suggestion: It seems like this might be slightly easier to read if this is inlined like:
```
{
    ATTR_ENTITY_ID: state.entity_id,
    **data
}
```

The motivation for the suggestion is there are just a few variables here to track (`service`, `data`, and `service_data`). However, what is here is perfectly correct, so consider this optional as it may be a style preference.